### DISCUSSION TITLE
Inso CLI lint spec with .spectral.yml error

### DISCUSSION THREAD

**Discussion by ponsakorn30214** [Kong/insomnia#7636] [Help] [answered]
I'm trying to use `inso lint spec` with a custom linting rule in `.spectral.yml` file.
However, using inso cli 9.3.1 causes errors:
```
$ inso lint spec petstore.yaml
 Load api specification with identifier petstore.yaml from data store
› Found 0.
› Linting specification from file /home/petstore.yaml
› Loading ruleset from .spectral.yml

 FATAL  Failed to read "/home/petstore.yaml"
```
I tested this setup with Insomnia 8.6.1 and it works normally.
Are there any update I'm missing on inso CLI 9.x?

**Comment by subnetmarco** [comment]
Prioritizing for 9.3.2 this week.

**Reply by ponsakorn30214** [reply]
Thank you Marco, Do you want me to file a bug issue on this?

**Reply by subnetmarco** [reply]
No thanks, we are already tracking it.

**Reply by jackkav** [reply]
@ponsakorn30214 Hi thanks for the report.
I've been looking at how to simplify the pathing configuration of inso so your feedback is much appreciated.
Could you provide some context to help me understand where your files are. 
Where is you petstore.yaml located?
Where is your inso executable located? 
Is it added to $PATH or do you run it from where you downloaded it to?
Do you typically use any other inso features besides lint?

**Reply by ponsakorn30214** [reply]
@jackkav happy to provide the info.
Both `.spectral.yml` and `petstore.yaml` are in my home directory of `/home/`
I've check the permission, all of the files are at least `600`, binary at `755` with all of them as my user as the owner.
I've tried using inso as a binary using `./inso` and putting it at `/usr/bin`. `/usr/bin` is in my PATH env var.

I've used inso cli since v3.0 both as a standalone and in combination with Insomnia to run test, export spec, etc..

**Reply by jackkav** [reply]
I found the root cause, the bug was introduced in #7203 when something (as yet undetermined) between spectral, node20 and esbuild module resolution causes the `bundleAndLoadRuleset` to throw the following exception, which is swallowed by inso.
```
Error [ERR_REQUIRE_ESM]: require() of ES Module /snapshot/insomnia/node_modules/jsep/dist/cjs/jsep.cjs.js from /snapshot/insomnia/node_modules/simple-eval/dist/index.js not supported.
jsep.cjs.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead either rename jsep.cjs.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /snapshot/insomnia/node_modules/jsep/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at Module.require (pkg/prelude/bootstrap.js:1851:31)
    at Object.<anonymous> (/snapshot/insomnia/node_modules/simple-eval/dist/index.js:5:12)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22) {
  code: 'ERR_REQUIRE_ESM'
}

```

I'm going to ensure errors like this don't get swallowed and keep trying to figure out what is breaking the bundler, but in the meantime you can use the [spectral-cli](https://github.com/stoplightio/spectral) directly.
`spectral lint petstore.yaml`

**Reply by jackkav** [reply]
I have an update for you. It would appear the spectral bundle ruleset module, searches the parent directory for ruleset imports even when you are using the "spectral:oas" ruleset. 

This explains why running this command from your /home directory is causing you issues. It seems spectral is expecting you to have full access to the parent folder of the file you are linting. 

I suspect theres something in the spectral docs to explain this behaviour.

Hope this helps @ponsakorn30214

**Reply by ponsakorn30214** [reply]
@jackkav Thank you for looking into this issue.
I still confused about this part:

> It seems spectral is expecting you to have full access to the parent folder of the file you are linting.

I've tried running `inso lint spec` somewhere else, e.g. `/home/user1/` but the same error occured.

> I suspect theres something in the spectral docs to explain this behaviour.

Noted, I will dig into spectral doc for this issue then.

Apart from that, I believe I have to wait for newer patch version of inso to be able to use linting the same way I've been using since v3.0 right?

**Reply by jackkav** [reply]
Can you try it with spectral cli directly, as I said we're just a wrapper. So I suspect you'll get the same issue with spectral but perhaps with better logging.

This case works fine in our tests run from within the insomnia git directory, but it fails when I run it in my home folder even when i create folders there too. I think `.spectral.js` is searching for an import of some kind but I'm not familiar with the code to know where to look.

I also found when changing the content extends of the .spectral.yml to [] this issue would go away, since we fallback to spectral:oas that could be a workaround

this is the error we get when spectral tries to read the `.spectral.yml`
```
stderr: '/Users/jack/deleteme/deleteme2' is imported by ../../../../deleteme/deleteme2/.spectral.js, but could not be resolved – treating it as an external dependency

stderr: q_ [Error]: Ruleset must have rules or extends or overrides defined
    at H5t (/snapshot/insomnia/packages/insomnia-inso/dist/index.js)
    at new e (/snapshot/insomnia/packages/insomnia-inso/dist/index.js)
    at /snapshot/insomnia/packages/insomnia-inso/dist/index.js
    at Array.reduce (<anonymous>)
    at new e (/snapshot/insomnia/packages/insomnia-inso/dist/index.js)
    at vkt (/snapshot/insomnia/packages/insomnia-inso/dist/index.js)
    at async fOe (/snapshot/insomnia/packages/insomnia-inso/dist/index.js)
    at async Promise.all (index 0) {
  code: 'invalid-ruleset-definition',
  path: []
}
```

**Reply by ponsakorn30214** [reply]
> Can you try it with spectral cli directly

I run `spectral lint petstore.yaml` and get the following result (spectral 6.11.1, maybe I can try your version) :
```
/home/petstore.yaml
  2:6   warning  info-contact           Info object must have "contact" object.                        info
  2:6   warning  info-description       Info "description" must be present and non-empty string.       info
  11:9  warning  operation-description  Operation "description" must be present and non-empty string.  paths./pets.get
 15:11  warning  operation-tag-defined  Operation tags must be defined in global tags.                 paths./pets.get.tags[0]
 43:10  warning  operation-description  Operation "description" must be present and non-empty string.  paths./pets.post
 47:11  warning  operation-tag-defined  Operation tags must be defined in global tags.                 paths./pets.post.tags[0]
  64:9  warning  operation-description  Operation "description" must be present and non-empty string.  paths./pets/{petId}.get
 68:11  warning  operation-tag-defined  Operation tags must be defined in global tags.                 paths./pets/{petId}.get.tags[0]

✖ 8 problems (0 errors, 8 warnings, 0 infos, 0 hints)
```
 
running `inso lint spec petstore.yaml` at the same place got me the `Failed to read "/home/petstore.yaml"` error.

> but it fails when I run it in my home folder even when i create folders there too.

yeah, I tested at different places and even run as root but got the same error.

> I also found when changing the content extends of the .spectral.yml to [] this issue would go away

Not for me tho, leaving `.spectral.yml` to just `extends: []` still give me the `Failed to read` error

**Reply by jackkav** [reply]
I have a suspicion it could be a pathing/external dependency issue during bundling since it continues to work as expected when bundled and packaged within the insomnia project directory. Perhaps its searching for a node_module in the parent folder? The method of importing spectral.yml with spectral code is causing the indirection and making debugging difficult here. I'll keep you updated if we find a solution.

**Reply by jackkav** [reply, marked as answer]
I found a workaround, it would appear that by installing a spectral node modules near the file you want to run it is able to work.
for example I have a spec and custom lint ruleset at path /home/git/project/spec.yml and /home/git/project/.spectral.yml
```
cd /home/git/project
npm i @stoplight/spectral-ruleset-bundler
inso lint spec /home/git/project/spec.yml
````
Will run successfully because the spectral ruleset bundler is expecting internal js files to be available near the .spectral.yml location, ignoring that this dependency is packaged with inso, perhaps because it is packaged with errors, the root cause is still unclear.

Related issue: https://github.com/stoplightio/spectral/issues/1956
